File: sonar-server/src/main/webapp/WEB-INF/app/models/filter_result.rb
 #
 # Sonar, entreprise quality control tool.
 # Copyright (C) 2009 SonarSource SA
 # mailto:contact AT sonarsource DOT com
 #
 # Sonar is free software; you can redistribute it and/or
 # modify it under the terms of the GNU Lesser General Public
 # License as published by the Free Software Foundation; either
 # version 3 of the License, or (at your option) any later version.
 #
 # Sonar is distributed in the hope that it will be useful,
 # but WITHOUT ANY WARRANTY; without even the implied warranty of
 # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 # Lesser General Public License for more details.
 #
 # You should have received a copy of the GNU Lesser General Public
 # License along with {library}; if not, write to the Free Software
 # Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
 #
 class FilterResult
  attr_accessor :page_size, :page_id, :security_exclusions, :filter

  def initialize(filter, options={})
    @filter = filter
    @page_size=options[:page_size] || @filter.page_size
    @page_id=(options[:page_id] ? options[:page_id].to_i : 1)
    @sids=options[:snapshot_ids]
    @page_sids=[]
    @security_exclusions=options[:security_exclusions]
    @metric_ids=(options[:metric_ids] || @filter.columns.map{|col| col.metric ? col.metric.id : nil}.compact.uniq)

    from=(@page_id-1) * @page_size
    to=(@page_id*@page_size)-1
    to=@sids.size-1 if to>=@sids.size

    @measures_by_snapshot={}
    @page_snapshots=[]
    @snapshots_by_id={}
    @links_by_pid={}
    if from<@sids.size
      #
      # load snapshots and resources
      #
      @page_sids=@sids[from..to]
      @page_snapshots=Snapshot.find(:all, :include => ['project'], :conditions => ['id in (?)', @page_sids])
      @page_snapshots.each{|s| @snapshots_by_id[s.id]=s}

      if @page_sids.size>0
        #
        # load measures
        #
        if @metric_ids.size>0
          measures=ProjectMeasure.find(:all, :conditions => ['rule_priority is null and rules_category_id is null and rule_id is null and snapshot_id in (?)', @page_sids])

          if filter.display_user_managed_metrics?
            measures.concat(AsyncMeasureSnapshot.search(@page_sids, @metric_ids))
          end

          measures.each do |m|
            snapshot=@snapshots_by_id[m.snapshot_id]
            @measures_by_snapshot[snapshot]||={}
            @measures_by_snapshot[snapshot][m.metric]=m
          end
        end

        #
        # load links
        #
        if @filter.display_links?
          pids=@page_snapshots.map{|snapshot| snapshot.project_id}
          ProjectLink.find(:all, :conditions => {:project_id => pids}, :order => 'link_type').each do |link|
            @links_by_pid[link.project_id] ||= []
            @links_by_pid[link.project_id]<<link
          end
        end
      end
    end
  end

  def size
    @sids.size
  end

  def empty?
    @page_sids.empty?
  end

  def page_count
    result=@sids.size / @page_size
    result+=1 if (@sids.size % @page_size > 0)
    result
  end

  def page_sorted_snapshot_ids
    @page_sids
  end

  def snapshots
    @page_snapshots
  end

  def snapshot(sid)
    @snapshots_by_id[sid]
  end

  def measure(snapshot, metric)
    if metric
      hash=@measures_by_snapshot[snapshot]
      hash ? hash[metric] : nil
    else
      nil
    end
  end

  def measures_by_snapshot
    @measures_by_snapshot
  end

  def links(resource_id)
    @links_by_pid[resource_id] || []
  end

  def security_exclusions?
    @security_exclusions==true
  end

 end

File: sonar-server/src/main/java/org/sonar/server/filters/FilterResult.java
/*
 * Sonar, open source software quality management tool.
 * Copyright (C) 2008-2012 SonarSource
 * mailto:contact AT sonarsource DOT com
 *
 * Sonar is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * Sonar is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Sonar; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
 */
package org.sonar.server.filters;

import com.google.common.collect.Ordering;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;

public class FilterResult {
  private List<Object[]> rows;
  private Filter filter;
  public static final int SORTED_COLUMN_INDEX = 3;

  public FilterResult(Filter filter, List<Object[]> rows) {
    this.rows = new ArrayList<Object[]>(rows);
    this.filter = filter;
  }

  /**
   * @return a list of arrays
   */
  public List<Object[]> getRows() {
    return rows;
  }

  public int size() {
    return rows.size();
  }

  public Integer getSnapshotId(Object[] row) {
    return (Integer) row[getSnapshotIdIndex()];
  }

  public Integer getProjectId(Object[] row) {
    return (Integer) row[getProjectIdIndex()];
  }

  public Integer getRootProjectId(Object[] row) {
    return (Integer) row[getRootProjectIdIndex()];
  }

  public int getSnapshotIdIndex() {
    return 0;
  }

  public int getProjectIdIndex() {
    return 1;
  }

  public int getRootProjectIdIndex() {
    return 2;
  }

  public void sort() {
    if (filter.isSorted()) {
      Comparator<Object[]> comparator = (filter.isTextSort() ? new StringIgnoreCaseComparator(SORTED_COLUMN_INDEX) : new NumericComparator(SORTED_COLUMN_INDEX));
      if (!filter.isAscendingSort()) {
        comparator = Ordering.from(comparator).reverse();
      }
      Collections.sort(rows, comparator);
    }
  }

  public void removeUnvalidRows() {
    int numberOfCriteria = filter.getMeasureCriteria().size();
    if (numberOfCriteria > 0) {
      int fromColumnIndex = (filter.isSorted() ? SORTED_COLUMN_INDEX + 1 : SORTED_COLUMN_INDEX);
      for (Iterator<Object[]> it = rows.iterator(); it.hasNext(); ) {
        Object[] row = it.next();
        boolean remove = false;
        for (int index = 0; index < numberOfCriteria; index++) {
          if (row[fromColumnIndex + index] == null) {
            remove = true;
          }
        }
        if (remove) {
          it.remove();
        }
      }
    }
  }

  static final class NumericComparator implements Comparator<Object[]>, Serializable {
    private static final long serialVersionUID = 4627704879575964978L;
    private int index;

    NumericComparator(int index) {
      this.index = index;
    }

    public int compare(Object[] a1, Object[] a2) {
      Comparable c1 = (Comparable) a1[index];
      Comparable o2 = (Comparable) a2[index];

      return (c1 == null ? -1 : (o2 == null ? 1 : c1.compareTo(o2)));
    }
  }

  static final class StringIgnoreCaseComparator implements Comparator<Object[]>, Serializable {
    private static final long serialVersionUID = 963926659201833101L;
    private int index;

    StringIgnoreCaseComparator(int index) {
      this.index = index;
    }

    public int compare(Object[] o1, Object[] o2) {
      String s1 = (String) o1[index];
      if (s1 == null) {
        return -1;
      }
      String s2 = (String) o2[index];
      if (s2 == null) {
        return 1;
      }
      return s1.compareToIgnoreCase(s2);
    }
  }
}

